### PR TITLE
Extend playbook validation: Error out if there is an unknown key in vars/ansible_bender/

### LINKED ISSUE #92: config validation: error out if there is an unknown key in vars/ansible_bender/
To prevent typos.

### PR DESCRIPTION
**Pull request by kmehant**
Fixes #92

### CODE REVIEW

File: ansible_bender/constants.py [hunk 1]
```diff
@@ -32,3 +32,30 @@
...
+        "ansible_extra_args",
+        "python_interpreter",
+        # required keys
+        "playbook_path",
+        "metadata",
+        "state",
+        "base_image",
+        "target_image",
+        "builder_name"
+]
```
Comment (TomasTomecek) [inline ansible_bender/constants.py]: these are the keys from the database, not vars -> ansible_bender

Comment (kmehant) [reply ansible_bender/constants.py]: @TomasTomecek Sorry! I will work on it

**Comment by kmehant** [comment]
@TomasTomecek I have updated the list, hope everything is covered for vars/ansible_bender

**Comment by kmehant** [comment]
@TomasTomecek for the multilevel extraction of keys in the dictionary, I am thinking of using a stack to solve it, what are your thoughts regarding it.

**Comment by TomasTomecek** [comment]
I looked at json schema if it's able to help us here and... It seems it is!

https://stackoverflow.com/questions/53770325/additionalproperties-rule-in-json-schema-is-not-applied-to-nested-level-proper

Seems like that all we need to do is to enable `additionalProperties: false` on all the levels which will disallow unknown keys in the dict

**Comment by kmehant** [comment]
> I looked at json schema if it's able to help us here and... It seems it is!
> 
> https://stackoverflow.com/questions/53770325/additionalproperties-rule-in-json-schema-is-not-applied-to-nested-level-proper
> 
> Seems like that all we need to do is to enable `additionalProperties: false` on all the levels which will disallow unknown keys in the dict

That will be an awesome fix :stars:, thank you

**Comment by kmehant** [comment]
\packit test

**Comment by kmehant** [comment]
> I looked at json schema if it's able to help us here and... It seems it is!
> 
> https://stackoverflow.com/questions/53770325/additionalproperties-rule-in-json-schema-is-not-applied-to-nested-level-proper
> 
> Seems like that all we need to do is to enable `additionalProperties: false` on all the levels which will disallow unknown keys in the dict

@TomasTomecek What I found was BUILD_SCHEMA has got only one properties key so I can add additionalProperties key only to one level.

File: ansible_bender/schema.py [hunk 2]
```diff
@@ -223,7 +223,8 @@
             "examples": [
                 "/usr/bin/python4"
             ]
-        }
+        },
+    "additionalProperties": False
```
Comment (TomasTomecek) [inline ansible_bender/schema.py]: this is the schema for the db file, users are not meant to change this one, hence we don't need to validate the file so striuctly. Please set it in the other schema, the defined below - for the vars section. Thank you!

**Comment by TomasTomecek** [comment]
Also a test case here would be awesome!

File: ansible_bender/schema.py (hunk 2 above)
Comment (kmehant) [reply ansible_bender/schema.py]: @TomasTomecek I was actually confused, thank you for helping me out.
I found that keys such as`verbose_layer_names` were not there in that schema.

**Comment by thrix** [comment]
> /packit test

The F30 copr issue should be fixed now hopefully:
https://gitlab.com/testing-farm/cruncher/merge_requests/7

**Comment by kmehant** [comment]
@TomasTomecek I am not sure where actually the problem lies, I request your review on this.

**Comment by TomasTomecek** [comment]
I finally figured it out, now I understand why you were so confused about this - took me some time to understand what's wrong. Here we go:

1. You were right - validate method of ImageMetadata is not being called anywhere:
https://github.com/ansible-community/ansible-bender/blob/1cf0518581fb0447c9fe48108b2f229bf6f6297b/ansible_bender/conf.py#L77
2. The code which needs to be changed lies in core.py, class PbVarsParser
https://github.com/ansible-community/ansible-bender/blob/1cf0518581fb0447c9fe48108b2f229bf6f6297b/ansible_bender/core.py#L235
3. That class creates a playbook which exapands all the variables defined in vars section of a playbook - this has to be done so that people can use jinja templating in ansible_bender section of vars
4. And that's place where we need to introduce validation.
5. So once we update our internal structure with data from the playbook, we should validate that our structure is valid, that would be here:
https://github.com/ansible-community/ansible-bender/blob/1cf0518581fb0447c9fe48108b2f229bf6f6297b/ansible_bender/core.py#L365

That's great you wrote a test case for this - we can now check whether this would be a better solution to the problem.

**Comment by kmehant** [comment]
> I finally figured it out, now I understand why you were so confused about this - took me some time to understand what's wrong. Here we go:
> 
> 1. You were right - validate method of ImageMetadata is not being called anywhere:
>    https://github.com/ansible-community/ansible-bender/blob/1cf0518581fb0447c9fe48108b2f229bf6f6297b/ansible_bender/conf.py#L77
> 2. The code which needs to be changed lies in core.py, class PbVarsParser
>    https://github.com/ansible-community/ansible-bender/blob/1cf0518581fb0447c9fe48108b2f229bf6f6297b/ansible_bender/core.py#L235
> 3. That class creates a playbook which exapands all the variables defined in vars section of a playbook - this has to be done so that people can use jinja templating in ansible_bender section of vars
> 4. And that's place where we need to introduce validation.
> 5. So once we update our internal structure with data from the playbook, we should validate that our structure is valid, that would be here:
>    https://github.com/ansible-community/ansible-bender/blob/1cf0518581fb0447c9fe48108b2f229bf6f6297b/ansible_bender/core.py#L365
> 
> That's great you wrote a test case for this - we can now check whether this would be a better solution to the problem.

@TomasTomecek I have worked out as you said, I wish to have your review, again!

**Comment by TomasTomecek** [comment]
> /packit test

no need to do that: every push is first built as an RPM and then used in the testing infra

**Comment by thrix** [comment]
meh, sorry for the delay, I was lazy to implement cleanup of jobs and this is the result :( Cleaning up manually now

**Comment by thrix** [comment]
things will resume soonish

**Comment by TomasTomecek** [comment]
okay so let's green this PR, merge it and release ab 0.8

**Pull request merged**